docs(xo-server): guidelines (#8583)
vatesfr/xen-orchestra: packages/xo-server/docs/guidelines.md

1. **Handling**  
   Prevents midnight outages from uncaught crashes.

2. **Async + Timeouts**  
   Avoids hung processes that require manual restart.

3. **Dependency Control**  
   Reduces security vulnerabilities and upgrade headaches.

4. **Readability**  
   Cuts onboarding time for new engineers.

5. **Scalability**  
   Ensures the system works with 50K VMs as well as 50.

6. **Resource Management**  
   Prevents memory leaks and corrupted data from unclosed resources.

---

## Core Principles

1. Works First Priority

   - ✅ Passes basic smoke tests before optimization
   - ❌ Rewrote the entire module to shave 0.1ms off runtime
   - ⚠️ Works but skipped testing due to time constraints

2. No Trivial Changes

   - ✅ Fixed typo in log message + verified no side effects
   - ❌ Changed 'status' to 'state' globally without grep-checking
   - ⚠️ "Refactored" variable names but introduced a shadowed variable

3. Improve Continuously
   - ✅ Added type hints while fixing a bug
   - ❌ "This code is ugly but I won't touch it"
   - ⚠️ Refactored without tests "because it looked simple"

---

## Key Rules

### Asynchronous Code

- any asynchonous code must be awaited

> - consider the use of a timeout handler

Author: ```suggestion
- consider the use of a timeout handler
- if the method returns a Promise , it does not need to be declared as async. Remember to use jsdoc to declare the return type as Promise <type> 

```

```ts
// ✅
try {
  db.insert();
} catch (error) {
  log.warn('DB full', { error });
  throw;
}
try{

}catch(err){
   const specificError= new Error('this is a specific error')
   specificError.cause = err
   specificError.code = 'ENOENT'
}

// ❌
JSON.parse(input); // No try/catch
try{
   asyncProcess() // no await => unbounded error
}catch(error){}
promise.catch(err=>throw err) // unbounded no stack trace

// ⚠️
try { parse(); } catch {} // Silenced error
try{

}catch(err){
   throw new Error('this is a specific error')// lost the original error
}
```

### Logging

Use structured logs with `@xen-orchestra/log` avoid flooding.

```ts
// ✅
log.info('VM started', { vmId, durationMs })

// ❌
console.log(`${vm} started at ${new Date()}`) // Unstructured

// ⚠️
log.error(err) // No context like VM ID
ayncPromise.catch(() => {}) // erro is forcibly silenced
```

### Readability

- name method and variable explicitly, especially on non typed code. For example, set/edit method should only be used on a typed object. Avoid generic code like for (const object of objects){}
- method parameter are readonly and explicit. They are typed.
- function name are readonly and explicit.

> - limit number of parameters or use oject parameter

Reviewer: ```suggestion
- limit number of parameters or use object parameter
```

```ts
// ✅
for (const vm of runningVms) {
  await vm.stop()
}
function work(ref, { bootable, seekable, focusable = true, otherboolean = false } = {}) // work(1, {bootable: true , focusable:false})
// ❌
for (const x of y) {
  /* What's x? */
}

let set = (item, key, value) => (item[key] = value)
set = item => set(item, 'mykey', 4)

// ⚠️
items.forEach(item => process(item)) // 'item' too vague
function work(ref, bootable, seekable, focusable, otherboolean) //work(1, true, true, false,true) or work(1, true, false, false,true)
```

### Scalability

No unbounded memory/parallel ops.

```ts
// ✅
for await (const chunk of stream) {
  process(chunk)
}

// ❌
const allData = await fs.readFile('50TB.log')

// ⚠️
await Promise.all(users.map(fetchDetails)) // 50K parallel requests
```

### Testing

Use `node:assert`; test critical paths.

```js
// ✅
assert.strictEqual(result.status, 'running')

// ❌
if (result !== 'running') throw Error() // Ad-hoc

// ⚠️
assert(result) // No error message
```

### Typing

> Zero `any`; JSDoc/TypeScript everywhere.

Author: ```suggestion
- Zero `any`; JSDoc/TypeScript everywhere.
- XO specific : any property change in `packages/xo-server/src/xapi-object-to-xo.mjs` must be reflected in `@vates/types`

```

> 

Author: add a part for JSdoc

```ts
// ✅
/** @type {Array<{id: string}>} */
const vms = []

// ❌
const data = {} // Untyped

// ⚠️
function parse(input: any) {} // Leaky abstraction
```

> Node 23 is around the corner and support a subset of typescript. It doesn't support enums, namespaces and parameters properties. Also is need to use the `type` keywork when importing type. If a module don't need this feature, it will be able to skip the compilation phase at build and keep te type checking in developpment mode

Reviewer 2: ```suggestion
Node 23 is around the corner and support a subset of typescript. It doesn't support enums, namespaces and parameters properties. Also is need to use the `type` keywork when importing type. If a module don't need this feature, it will be able to skip the compilation phase at build and keep the type checking in developpment mode
```

> Node 23 is around the corner and support a subset of typescript. It doesn't support enums, namespaces and parameters properties. Also is need to use the `type` keywork when importing type. If a module don't need this feature, it will be able to skip the compilation phase at build and keep the type checking in developpment mode

Author: ```suggestion
Node 23 supports a subset of typescript. It doesn't support enums, namespaces and parameters properties. Also is need to use the `type` keywork when importing type. If a module don't need this feature, it will be able to skip the compilation phase at build and keep the type checking in developpment mode
```

# Backend Project Guidelines

> Theses guidelines are not absolute rules, but expect to have to explain to the reviewer if the PR does not explicit why the guideline is not used.

Reviewer 3: ```suggestion
These guidelines are not absolute rules, but expect to have to explain to the reviewer if the PR does not explicitly state why the guideline is not used.
```

Reviewer 4: ```suggestion
These guidelines are not absolute rules, but if they are purposefully ignored, a justification will be needed for the reviewer.
```

Had to re-read this line 3 or 4 times to understand it.

```

### Error Handling

- error must be caught if they are recoverable
- don't silence error. Generally use a warn on unexpected error, an info on an expected error
- use error.cause or decorate original error with the relevant context.
- consider adding the current stacktrace to a new Error
- add an error code or a custom class extending error for custom errors

```ts
// ✅
try {
  db.insert();
} catch (error) {
  log.warn('DB full', { error });
  throw;
}
try{

}catch(err){
   const specificError= new Error('this is a specific error')
   specificError.cause = err
   specificError.code = 'ENOENT'
}

// ❌
JSON.parse(input); // No try/catch
try{
   asyncProcess() // no await => unbounded error
}catch(error){}
promise.catch(err=>throw err) // unbounded no stack trace

// ⚠️
try { parse(); } catch {} // Silenced error
try{

}catch(err){
   throw new Error('this is a specific error')// lost the original error
}
```

### Logging

Use structured logs with `@xen-orchestra/log` avoid flooding.

```ts
// ✅
log.info('VM started', { vmId, durationMs })

// ❌
console.log(`${vm} started at ${new Date()}`) // Unstructured

// ⚠️
log.error(err) // No context like VM ID

> ayncPromise.catch(() => {}) // erro is forcibly silenced

Reviewer: ```suggestion
ayncPromise.catch(() => {}) // error is forcibly silenced
```

- method parameter are readonly and explicit. They are typed.
- function name are readonly and explicit.
- limit number of parameters or use object parameter

```ts
// ✅
for (const vm of runningVms) {
  await vm.stop()
}
function work(ref, { bootable, seekable, focusable = true, otherboolean = false } = {}) // work(1, {bootable: true , focusable:false})
// ❌
for (const x of y) {
  /* What's x? */
}

let set = (item, key, value) => (item[key] = value)
set = item => set(item, 'mykey', 4)

// ⚠️
items.forEach(item => process(item)) // 'item' too vague
function work(ref, bootable, seekable, focusable, otherboolean) //work(1, true, true, false,true) or work(1, true, false, false,true)
```

### Scalability

No unbounded memory/parallel ops.

```ts
// ✅
for await (const chunk of stream) {
  process(chunk)
}

// ❌
const allData = await fs.readFile('50TB.log')

// ⚠️
await Promise.all(users.map(fetchDetails)) // 50K parallel requests
```

### Testing

Use `node:assert`; test critical paths.

```js
// ✅
assert.strictEqual(result.status, 'running')

// ❌
if (result !== 'running') throw Error() // Ad-hoc

// ⚠️
assert(result) // No error message
```

### Typing

> - Zero `any`; JSDoc/TypeScript everywhere.

Author: ```suggestion
- Zero `any`; JSDoc/TypeScript everywhere. https://jsdoc.app/about-getting-started  . JSDoc offer capabilties to define and import types, and `@implements` to handle interface  
```

- XO specific : any property change in `packages/xo-server/src/xapi-object-to-xo.mjs` must be reflected in `@vates/types`

> 

Author: ```suggestion
/**
```

> - Zero `any`; JSDoc/TypeScript everywhere. https://jsdoc.app/about-getting-started  . JSDoc offer capabilties to define and import types, and `@implements` to handle interface  

Reviewer: ```suggestion
- Zero `any`; JSDoc/TypeScript everywhere. https://jsdoc.app/about-getting-started  . JSDoc offers capabilties to define and import types, and `@implements` to handle interface  
```
